ramips: further improve support for Xiaomi MiWiFi Nano

mans0n opened the pull request:
This patch does the following:

- prepend vendor name to model
- set status LEDs to follow the behavior in stock FW
- add label mac address
- simplify state_default node definition
- fix switch setup
- remove redundant ralink,mtd-eeprom
- use generic name for flash node

Stock FW status indicators:
https://files.xiaomi-mi.com/files/Mi_Router_Wi-Fi_Nano/Mi_router-NANO_EN.pdf
> Yellow: power on / off
> Blue: during normal operation
> Red: in case of problems with the operation of the device

MiWiFi Nano has two LAN ports, which are in reverse order. Add port numbers
to them, and disable unused ports.

mt76 does not use ralink,mtd-eeprom, and mediatek,mtd-eeprom is already set
in SoC DTSI. mt76 will read wmac MAC address from offset +4 automatically.

mans0n committed:
ramips: further improve support for Xiaomi MiWiFi Nano

This patch does the following:

- prepend vendor name to model
- set status LEDs to follow the behavior in stock FW
- simplify state_default node definition
- use generic name for flash node

Stock FW status indicators:
https://files.xiaomi-mi.com/files/Mi_Router_Wi-Fi_Nano/Mi_router-NANO_EN.pdf
> Yellow: power on / off
> Blue: during normal operation
> Red: in case of problems with the operation of the device

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Some notes and questions:
1. MiWiFi Nano actually uses MT7628NN. Would it be better to change SOC?
2. Mi Router 4A 100M edition has DTS similar to that of MiWiFi Nano, but I don't feel they are similar enough to have shared DTSI.
3. Actually WAN MAC address is same as LAN MAC address. Is it okay to share the same address between LAN and WAN?

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-nano.dts:
```diff
@@ -67,9 +66,13 @@
+
 &wmac {
 	status = "okay";
-	ralink,mtd-eeprom = <&factory 0x4>;
```
This is not "mtd-mac-address", but "mtd-eeprom"!
So, is this really ignored? I calculated MAC addresses from that (using 0x8) which would be wrong then, too.

mans0n replied:
It is added by b28e94d4bfa1. I'm not sure about the author's intention...
Anyway, ralink,* is only used in rt2x00, and mt76 only uses mediatek,*.
```
$ grep -rIn "ralink,"
drivers/net/wireless/ralink/rt2x00/rt2x00dev.c:1022:    if (!of_property_read_u32(np, "ralink,2ghz",
drivers/net/wireless/ralink/rt2x00/rt2x00dev.c:1025:    if (!of_property_read_u32(np, "ralink,5ghz",
drivers/net/wireless/ralink/rt2x00/rt2800soc.c:227:     { .compatible = "ralink,rt2880-wmac" },
drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c:49:   list = of_get_property(np, "ralink,mtd-eeprom", &size);
drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c:82:   if (of_find_property(np, "ralink,mtd-eeprom-swap", NULL))
drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c:109:  if (np && of_property_read_string(np, "ralink,eeprom", &eep) == 0)
drivers/net/wireless/ralink/rt2x00/rt2800lib.c:11183:           if (!of_property_read_u32(np, "ralink,led-polarity",

$ grep -rIn "mediatek,"
drivers/net/wireless/mediatek/mt76/mt7603/soc.c:67:     { .compatible = "mediatek,mt7628-wmac" },
drivers/net/wireless/mediatek/mt76/eeprom.c:29:         list = of_get_property(np, "mediatek,mtd-eeprom", &size);
```

adrianschmutzler replied:
Ah, thanks for the research. Indeed, the commit message talks of MAC addresses. Do you own the device, and may just quickly check whether the address is in 0x4 or 0x8?

mans0n replied:
It's in 0x4.
`0040000  28 76 00 02 f0 b4 29 xx  xx xx 00 00 00 00 00 00  |(v....).........|`

adrianschmutzler replied:
Okay, so you are right then and I will look for the similar cases I remember.

adrianschmutzler replied:
https://patchwork.ozlabs.org/patch/1214746/
This will remove all cases (seems like quite a lot of copy/paste happened). Thanks for finally solving this odd thing.
I'd actually remove this change from the patchset here, as with my patch we have a separate commint as well as a more extensive explanation and your patch provides enough other valuable changes anyway.

mans0n replied:
Great! I've never noticed that... Thanks for your work. :)

adrianschmutzler replied:
I actually stumbled over this when changing label_mac in 02_network from (eth0+X) to flash locations in the first place (setting those strange 0x8 offsets). After all, even if nobody uses the label MAC address, having implemented it revealed so many errors in the existing setup that it was worth it just for that :-)

adrianschmutzler commented:
> Actually WAN MAC address is same as LAN MAC address. Is it okay to share the same address between LAN and WAN?

If that's the case on stock FW, yes, please change it here as well (in a separate patch, maybe include label MAC address there). There are some devices that use the same address for lan/wan/2.4 GHz in stock firmware, so this seems to be okay.

mans0n committed:
ramips: fix MAC address setup for Xiaomi MiWiFi Nano

MAC addresses are stored in factory partition at:
0x0004: WiFi 2.4GHz (label_mac +1)
0x0028: LAN, WAN (label_mac)

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Made a separate patch for MAC addresses.

adrianschmutzler commented:
Sorry, but I'd be happy if you also separated the switch/port changes into a separate commit. I will have to do research on how esw is set up correctly, and won't have time for that right away. I would then directly merge the remainder of patch 1 (without the factory 0x4 changes) and patch 2 (MAC address setup).

mans0n committed:
ramips: fix switch setup for Xiaomi MiWiFi Nano

MiWiFi Nano has two LAN ports, which are in reverse order. Add port numbers
to them, and disable unused ports.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

adrianschmutzler commented:
> 1. MiWiFi Nano actually uses MT7628NN. Would it be better to change SOC?

I'm not really sure about that. Maybe ask at the mailing list, but maybe nobody cares there.

> 2. Mi Router 4A 100M edition has DTS similar to that of MiWiFi Nano, but I don't feel they are similar enough to have shared DTSI.

Wouldn't do that either.

Do you own a Mi Router 4A 100M? Would be interested in a hexdump of factory there, as all addresses are currently derived from 0x4. Despite, one could maybe raise flash frequency.

mans0n commented:
Removed ralink,mtd-eeprom part from the patch, and made a separate patch for switch setup.

mans0n commented:
> Do you own a Mi Router 4A 100M?

I don't. So I cannot make change to DTS of Mi Router 4A 100M by myself.

adrianschmutzler commented:
Okay, pushed 1 and 2.

adrianschmutzler commented:
Merged patch 3, so this is dealt with.
Thanks.

Neustradamus commented:
Merged commit is:
- https://github.com/openwrt/openwrt/commit/74e9f95c79dc574c2407ed798f37c9a4c989e55d
